File: src/fensu/config/main/resolve_target_root.py
"""Resolve a selected target root inside its repository."""

from pathlib import Path

from fensu.config.constants import DEFAULT_TARGET_ROOT
from fensu.config.exceptions import ConfigError
from fensu.config.models import Config, ResolvedTargetRoot


def resolve_target_root(*, config: Config, repo_root: Path) -> ResolvedTargetRoot:
    """Return one canonical path and repository-relative target identity."""

    repository: Path = repo_root.resolve()
    target: Path = (repository / config.target_root).resolve()
    if not target.is_relative_to(repository):
        raise ConfigError(
            f"Target {config.target or '<legacy>'} root {config.target_root!r} "
            "must not escape the repository."
        )
    relative: Path = target.relative_to(repository)
    repository_relative: str = DEFAULT_TARGET_ROOT if not relative.parts else relative.as_posix()
    return ResolvedTargetRoot(path=target, repository_relative=repository_relative)

File: crates/fensu-cli/src/configuration/main/resolve_target_root.rs
//! Resolve one validated target root from its repository boundary.

use std::path::{Path, PathBuf};

pub(crate) fn resolve_target_root(
    repository_root: &Path,
    configured: &str,
) -> Result<PathBuf, String> {
    crate::configuration::_helpers::roots::resolve_target_root(repository_root, configured)
}
